WalletWasabi/Blockchain/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -273,6 +273,33 @@ private ProcessedResult ProcessNoLock(SmartTransaction tx)
 			TransactionStore.AddOrUpdate(tx);
 		}
 
+		if (tx.Confirmed)
+		{
+			// If a key is internal and spent all its coins, then it shouldn't be used again.
+			foreach (var spenderKey in myInputs.Select(x => x.HdPubKey).Where(x => x.IsInternal).Distinct())
+			{
+				if (spenderKey.Coins.Any(x => !x.IsSpent()))
+				{
+					// The key still has unspent coins.
+					continue;
+				}
+
+				if (spenderKey.KeyState != KeyState.Spent)
+				{
+					KeyManager.SetKeyState(KeyState.Spent, spenderKey);
+					spenderKey.LatestSpentHeight = tx.Height;
+				}
+				else
+				{
+					if (spenderKey.LatestSpentHeight < tx.Height)
+					{
+						// We should never reach this point.
```

Comment:
Maybe add a warning log message here that says we shouldn't have reached this point and there might be a bug or something.

Reply:
It is not a bug, just somehow the user got money onto an internal address and spent it. I agree to add a **warning** log message that something strange is happening.